operator-common/src/main/java/io/strimzi/operator/common/operator/resource/DeploymentOperator.java

```diff
@@ -105,6 +105,12 @@ protected Integer currentScale(String namespace, String name) {
         return waitFor(reconciliation, namespace, name, "observed", pollIntervalMs, timeoutMs, this::isObserved);
     }
 
+    public Future<Void> readiness(Reconciliation reconciliation, String namespace, String name, long pollIntervalMs, long timeoutMs) {
```

Comment:
`AbstractReadyResourceOperator#isReady` has the javadoc `Check if a resource is in the Ready state.`, so the implied contract for `AbstractReadyResourceOperator#readiness` is "Wait until the resource is in the Ready state.". By overriding readiness here we're implicitly changing that contract for Deployments to be something slightly different, i.e. `status==Ready && metadata.generation==status.observedGeneration`. That could be confusing. 

I'm not sure if there are any places where we might want to wait for `status==Ready` but not `metadata.generation==status.observedGeneration`. If so, perhaps we should make the `&&` version the default semantic for `readiness()` (possibly moving `waitForObserved` up the class hierarchy, though I'm not sure all of the subclasses of ARO have a `status.observedGeneration`)?

In any case, I think we should try to define the contract for readiness as part of this PR.

Reply:
I think this is unique to Deployments as it controls the Deployment rollout basically.

Reply:
So @tombentley basically you mean that we should have the waitForObserve() method from inside of readiness in AbstractReadyResourceOperator? Since I guess

Reply:
But yeah at the same time I guess `waitforObserve` / metadata.generation==status.observedGeneration  is common for deployment ?

Reply:
I think our `readiness()` should have a well-defined contract so that the returned Future only returns when we know that the resource is ready and that that readiness pertains to the current generation of the resource.

Our `readiness()` is defined on `AbstractReadyResourceOperator`, which is a superclass of `PodOperator`, `StatefulSetOperator` and `DeploymentConfigOperator`, the resources for which all have statuses (I'm assuming with a `observedGeneration`, but I've not checked), but also the `EndpointOperator` where `Endpoints` doesn't have a `status`. While we can use the `status.observedGeneration` to check that the readiness reported by `Fabric8` (defined by `io.fabric8.kubernetes.client.internal.readiness.Readiness#isResourceReady`)  pertains to the current generation of the resource we can't do that for `Endpoints`. Thus I don't see how you can avoid a race when checking for `Endpoints` readiness.

I think it would be weird to special case `Deployments` as requiring the generation check but not the other status-having resources. Why is it OK for them to have a racy semantic but not Deployments?

Reply:
It is that the resources have different semantics and at the end their readiness means something slightly different in each case I would say. At the end, Fabric8 already implements it differently for every resource. Also the way the resources apply their updates is different. So I do not think this can be done the same way for all of them.

But I'm fine if you want to keep our `readiness` method clean and do always the same. But in that case I would suggest to close this PR and leave this as it is in main with separate methods which need to be called both instead of trying to force the Deployment rollout mechanism onto other resources.

Reply:
@scholzj I agree on keeping the seperate methods. So should I close this PR?

Reply:
Well, if @tombentley thinks this breaks the intention of the `readiness()` method then I would say yes, we should keep it as it is and close this and the related issues.

Reply:
I think it can be argued that Fabric8's readiness checks should take into account `status.observedGeneration` (where present on the resource). I've opened https://github.com/fabric8io/kubernetes-client/issues/3474 about this. Let's see what they say before closing this issue.